framework/core/js/tests/unit/common/utils/Translator.test.ts

```diff
@@ -0,0 +1,59 @@
+import Translator from '../../../../src/common/Translator';
+import extractText from '../../../../src/common/utils/extractText';
+
+/*
+ * These tests should be in sync with PHP tests in `tests/unit/Locale/TranslatorTest.php`, to make sure that JS
+ * translator works in the same way as JS translator.
+ */
+
+test('placeholders encoding', () => {
+  const translator = new Translator();
+  translator.addTranslations({
+    test1: 'test1 {placeholder} test1',
+    test2: 'test2 {placeholder} test2',
+  });
+
+  expect(extractText(translator.trans('test1', { placeholder: "'" }))).toBe("test1 ' test1");
+  expect(extractText(translator.trans('test1', { placeholder: translator.trans('test2', { placeholder: "'" }) }))).toBe("test1 test2 ' test2 test1");
+});
+
+test('missing placeholders', () => {
+  const translator = new Translator();
+  translator.addTranslations({
+    test1: 'test1 {placeholder} test1',
+  });
+
+  expect(extractText(translator.trans('test1', {}))).toBe('test1 {placeholder} test1');
+});
```

Comment:
@rob006 the only failing test is this one, isn't the test actually incorrect since the placeholder is neither escaped nor provided?

![image](https://github.com/user-attachments/assets/e35d4d52-ea18-485d-8b85-56eb0c085c00)

Reply:
Unfortunately, I don't remember the details. I think it was defined in this way, because PHP translator works like that, so the goal was to make JS and PHP implementations consistent. But I don't think this scenario is important and some discrepancy may be accepted in edge cases like that. I would just leave commented version of this test with note that this case works differently in PHP.

Reply:
Also, IIRC some JS implementations throws exception in case of missing params, so clearly there is no consensus what to do in that scenario. IMO leaving params as is is the golden mean - it does not break app completely, but it is clearly visible that something is missing.

Reply:
Makes sense. The only alternative thing we can do is leave it as undefined (I made a change to empty it) that does communicate that there should be something there, wdyt?

![Screenshot from 2024-10-23 18-14-44](https://github.com/user-attachments/assets/db239113-48fc-4d02-964b-37308ca467f8)

Reply:
I'm not sure, it is the 4th way of handling this and it may create confusing results. :D But it also kinda make sense in JS world.

Reply:
Maybe use `{undefined}` as placeholder in that case? `undefined` may sound like actual value in many cases and using it may make it even harder to realize that param is missing.